core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java

```diff
@@ -934,7 +968,21 @@ private RexNode flattenComparison(
         return conjunction;
       }
     }
+
+  }
+
+  private int getNewInnerOrdinal(RexNode firstOp, String literalString) {
+    int newInnerOrdinal = 0;
+    for (RelDataTypeField field : firstOp.getType().getFieldList()) {
+      if (literalString.equalsIgnoreCase(field.getName())) {
+        break;
```

Comment:
I don't think this is right to match the fist flattened field just by name, how do you know what the (struct type field original) operator is ? It seems only right if they are all `ITEM` operators.

Reply:
Yes, the method only used for item with string second operand.

Reply:
That means this code snippet only works for full `ITEM` operator, but there is no any code can guarantee this, the most weird part is that you match the `first` field with name, that confused me a lot, why not the second ? the third ?

Reply:
1) Every call to the method is performed only after the check. 
```
      if (rexCall.op == SqlStdOperatorTable.ITEM
          && rexCall.operands.get(0).getType().isStruct()
          && rexCall.operands.get(1).isA(SqlKind.LITERAL)
          && SqlTypeUtil.inCharFamily(rexCall.operands.get(1).getType())) {
```
So I think it's enough to guarantee that we're looking for inner field within struct returned by first operand of ITEM.   

2) What's actually wrong with matching of first field if iteration is performed for non flattened struct fields ? There is no threat that some nested field with same name will come up and match before the desired one. Maybe I misunderstand you and you're just worried about case insensitive match, then in my opinion it's unlikely that users will create structs, like ```STRUCT <aaa:INT, AAA:STRING>``` .

Reply:
1. You indeed do the check:
```java
      if (rexCall.op == SqlStdOperatorTable.ITEM
          && rexCall.operands.get(0).getType().isStruct()
          && rexCall.operands.get(1).isA(SqlKind.LITERAL)
          && SqlTypeUtil.inCharFamily(rexCall.operands.get(1).getType())) {
```
before calling this method, but within the method, you still have a recursive logic to flatten all the struct fields, which i think has already exceed the limitation of the check.

2. Assume we have an expression a(b(c, 'k1'), 'k2')['k1'] (a, b, c all return struct types), is it really right we return the field of field c['k1'] ? And how do you know that a, b, and c are all `ITEM` operators ?

Reply:
1. There is no further recursive calls within the ```getNewInnerOrdinal(RexNode firstOp, String literalString)``` method. Only recursion is applied to predict post flatten size of field which resides before desired one, and this ```postFlattenSize(field.getType())``` prediction doesn't require any matching.

Reply:
2. In the absence of recursive search for non item operands, the method will look for ```'k1'``` inside result type of ```a(b(c, 'k1'), 'k2')```, note that type is not flattened at the moment of search.

Reply:
The method `getNewInnerOrdinal` is invoked after the recursive invocation `flattenProjection`, why do you think the type is not flattened ?

Reply:
Recursive ```flattenProjection``` call in the case will just fill list ```firstOpFlattenedExps``` and don't change state of object ```firstOp``` .